### ISSUE TITLE
tmux pannel lines rendered wrongly when locale set to zh_CN

### ISSUE THREAD

**Issue by BlurryLight** [mintty/wsltty#259]
```
# To use common configuration in %APPDATA%\mintty, simply remove this file
ThemeFile=
Locale=C
Language=en_US
Term=xterm-256color
Font=Fixedsys
FontHeight=14
CursorType=line
ForegroundColour=0,255,255
Columns=120
CursorColour=128,0,64
BellType=0
BoldAsColour=yes
```
when Locale is set to **C**, the panel lines look fine. 
![图片](https://user-images.githubusercontent.com/14940338/94553543-40743780-028b-11eb-80ea-7cb65d537af9.png)

After I set the locale to **zh_CN**, which is 
```
# To use common configuration in %APPDATA%\mintty, simply remove this file
ThemeFile=
Locale=zh_CN
Language=en_US
Term=xterm-256color
Font=Fixedsys
FontHeight=14
CursorType=line
ForegroundColour=0,255,255
Columns=120
CursorColour=128,0,64
BellType=0
BoldAsColour=yes
```
the tmux panel lines is rendered like 
![图片](https://user-images.githubusercontent.com/14940338/94553862-b7a9cb80-028b-11eb-809c-c2fdd9c1877a.png)

The lines looks weired and some chars have been overlapped by the lines.

My wsl is Ubuntu 2004 and the wsltty is 3.3.0

**Comment by Biswa96** [comment]
* Does this happen if you run wsl.exe in Command Prompt or Powershell?
* Does chaging font type and size make any change to this issue?

**Comment by BlurryLight** [comment]
1. I guess the locale entry in the mintty menu only affects mintty, is that right?  I have tried to run tmux in cmd, powershll and mintty, only wsltty behaved weirdly on lines.

2. I tried serveral font types and sizes, including `consolas`, `fixedsys` and some other fonts, but it was useless.

**Comment by Biswa96** [comment]
Would you like to test this? If you have root access in your WSL distribution, 1. backup the `/etc/default/locale` file 2. create a new `/etc/default/locale` file and add this `LANG=zh_CN@cjknarrow`. 3. re-execute wsltty.

**Comment by Biswa96** [comment]
Would you like to test this? If you have root access in your WSL distribution,

1. backup the `/etc/default/locale` file
2. create a new `/etc/default/locale` file and add this `LANG=zh_CN@cjknarrow`
3. re-execute wsltty.

**Comment by BlurryLight** [comment]
The lines now render correctly but another problem appears: The tty cannot neither display cjk characters nor emojis.

![图片](https://user-images.githubusercontent.com/14940338/94561215-4cb1c200-0296-11eb-9a4c-9fb7fc5c5547.png)

my current mintty config is
```
# To use common configuration in %APPDATA%\mintty, simply remove this file
ThemeFile=flat-ui
Locale=zh_CN
Language=en_US
Term=xterm-256color
Font=Sarasa Term SC
FontHeight=14
CursorType=block
ForegroundColour=0,255,255
Columns=120
CursorColour=128,0,64
BellType=0
BoldAsColour=yes
Charset=UTF-8
```

**Comment by Biswa96** [comment]
Another way is to pass `-e LANG=$LANG` with wslbridge2 command inside mintty. /cc @mintty

**Comment by Biswa96** [comment]
Another way is to pass `-e LANG` with wslbridge2 command inside mintty. /cc @mintty

**Comment by mintty** [comment]
Manipulating locale is a tricky issue which requires some background know-how.
CJK locales trigger ambiguous-wide character width, which affects the width of the horizontal line.
And in fact, if your WSL shell does not have the same idea about the locale as the terminal, things will go very wrong.

**Comment by mintty** [comment]
Did you actually check the locale set in the WSL shell? I suppose it's the default C.UTF-8. Try to set `LC_ALL=zh_CN.UTF-8 bash`. You may have to install the locale first.
If you still think there's something to analyse, please remind me of how to create a tmux panel.

**Comment by BlurryLight** [comment]
It's somewhat confusing, at least for me,  about the **locale** word in the context. From my understanding, there are two **locales** in wsltty and wsl. One is the locale setting in the wsl ubuntu, which is the *LC_xx =XXX*, and another one is the **locale** in the mintty menu `Options->Text->locale`. I have no idea if the two are correlated or not but I guess they are not. 

So in my environment, my **Ubuntu locales** are set to `zh_CN.UTF-8`, which is set by `sudo update-locale LANG=zh_CN.UTF-8`. 
And the problem appears when I change the **mintty locale** .

When the **mintty locale** is set to zh_CN, the lines display wrongly.
![QQ截图20200930220314](https://user-images.githubusercontent.com/14940338/94696986-3d9c4400-036a-11eb-80f3-f90209e271af.png)

And it works well when the **mintty locale** is **C**.
![20200930221436](https://user-images.githubusercontent.com/14940338/94697191-79370e00-036a-11eb-9a94-bfb4cb3f85a6.png)

The font is `Fixedsys, 12pt`, and the Ubuntu Locale keeps `zh_CN.UTF-8`.  Everytime I change the **mintty locale**, I open a new mintty and create a new tmux instance to make sure the change makes effect.

**Comment by mintty** [comment]
There are two problems:
1. The locale isn't propagated to WSL. Actually you handled this properly by setting both locales consistently.
2. Locales are not identical in cygwin and WSL. Particularly, CJK locales have the "ambiguous wide" property in cygwin (the mintty environment) but they are "ambiguous narrow" in WSL. Meaning that a range of characters, traditionally included in CJK character sets for compatibility, are double-cell wide or not. This includes the horizontal line. As a workaround, cygwin introduced the locale suffix @cjknarrow, so setting LC_ALL=zh_CN.utf8@cjknarrow before starting mintty would make locales consistent and fix the problem. Unfortunately, there is no mechanism right now to set such a locale in the mintty configuration, you could only do it externally, from a command line. This is because I considered the explicit options Locale and Charset a legacy feature, but maybe they should be tweaked.

**Comment by mintty** [comment]
> When the mintty locale is set to zh_CN, the lines display wrongly.

Actually, as the mintty locale is not propagated towards WSL (where you have to set it explicitly), there is no benefit in setting this option for wsltty. I will add some documentation about this.
On the other hand, I am also considering to disable CJK ambiguous-wide character width in WSL mode.

**Comment by mintty** [comment]
Actually, while checking whether interworking of mintty with WSL could be enhanced on this issue, I cannot reproduce it.
While setting LANG=zh_CN.utf8 before starting mintty (with no related setting in the config file) triggers the wide line,
`mintty -o Locale=zh_CN -o Charset=UTF-8` does not, which should be sufficient to faciliate seamless running of WSL.

**Comment by mintty** [comment]
@Biswa96, I guess I should add `-e LANG -e LC_CTYPE -e LC_ALL` to the invocation of wslbridge2 from mintty.

**Comment by mintty** [comment]
> I cannot reproduce it.

Comment withdrawn. I'll tweak locale setup for a workaround.

**Comment by Biswa96** [comment]
I had said that workaround first. Though it works I am not fully satisfied with it.

**Comment by mintty** [comment]
Actually, I think this should be tweaked in cygwin rather than mintty.
I.e., assuming that Linux systems / libraries (glibc, musl?) do not apply ambiguous-wide behaviour even for jp, ko, zh locales if they are combined with UTF-8 character encoding. At least that's what the locale zh_CN.utf8 on WSL Ubuntu suggests. Does anyone know whether this is consistent with other Linux distributions?

**Comment by mintty** [comment]
Released 3.4.1.

**Comment by mintty** [comment]
Released 3.4.3 with further tweaks.
